flair/nn/decoder.py

```diff
@@ -121,3 +123,112 @@ def forward(self, embedded):
+        super(LabelVerbalizerDecoder, self).__init__()
+        if decoding not in ["dot-product", "cosine-similarity"]:
+            raise RuntimeError("Decoding method needs to be one of the following: dot-product, cosine-similarity")
+        self.label_encoder = label_encoder
+        self.verbalized_labels = self.verbalize_labels(label_dictionary)
+        self.decoding = decoding
+        if self.decoding == "cosine-similarity":
+            self.distance = CosineDistance()
+            self.distance_score_transformation = torch.nn.Identity()
+        self.to(flair.device)
+
+    @staticmethod
+    def verbalize_labels(label_dictionary: Dictionary) -> List[Sentence]:
+        """Takes a label dictionary and returns a list of sentences with verbalized labels.
+
+        Args:
+            label_dictionary (flair.data.Dictionary): The label dictionary to verbalize.
+
+        Returns:
+            A list of sentences with verbalized labels.
+
+        Examples:
+            label_dictionary = corpus.make_label_dictionary("ner")
+            verbalized_labels = LabelVerbalizerDecoder.verbalize_labels(label_dictionary)
+            print(verbalized_labels)
+            [Sentence: "begin person", Sentence: "inside person", Sentence: "end person", Sentence: "single org", ...]
+        """
+        verbalized_labels = []
+        for byte_label, idx in label_dictionary.item2idx.items():
+            str_label = byte_label.decode("utf-8")
+            if label_dictionary.span_labels:
+                if str_label == "O":
+                    verbalized_labels.append("outside")
+                elif str_label.startswith("B-"):
+                    verbalized_labels.append("begin " + str_label.split("-")[1])
+                elif str_label.startswith("I-"):
+                    verbalized_labels.append("inside " + str_label.split("-")[1])
+                elif str_label.startswith("E-"):
+                    verbalized_labels.append("ending " + str_label.split("-")[1])
+                elif str_label.startswith("S-"):
+                    verbalized_labels.append("single " + str_label.split("-")[1])
+            else:
+                verbalized_labels.append(str_label)
+        return list(map(Sentence, verbalized_labels))
+
+    def forward(self, inputs: torch.Tensor) -> torch.Tensor:
+        """Forward pass of the label verbalizer decoder.
+
+        Args:
+            inputs (torch.Tensor): The input tensor.
+
+        Returns:
+            The scores of the decoder.
+
+        Raises:
+            RuntimeError: If an unknown decoding type is specified.
+        """
+        if self.training or not self.label_encoder._everything_embedded(self.verbalized_labels):
+            self.label_encoder.embed(self.verbalized_labels)
+
+        label_tensor = torch.stack([label.get_embedding() for label in self.verbalized_labels])
+
+        if self.training or not self.label_encoder._everything_embedded(self.verbalized_labels):
```

Comment:
The second condition is not needed: During, training always store embeddings. Otherwise, do not.

Reply:
how should the decoder know about the embeddings_storage_mode? Trainer stores embeddings depending on the mode, thus we would need to include this param into the forward loss called in trainer.train().